experimental/transliterate/src/compile/mod.rs

```diff
@@ -139,41 +223,367 @@ impl RuleBasedTransliterator<'static> {
+        + DataProvider<IdStartV1Marker>
+        + DataProvider<IdeographicV1Marker>
+        + DataProvider<JoinControlV1Marker>
+        + DataProvider<LogicalOrderExceptionV1Marker>
+        + DataProvider<LowercaseV1Marker>
+        + DataProvider<MathV1Marker>
+        + DataProvider<NoncharacterCodePointV1Marker>
+        + DataProvider<PatternSyntaxV1Marker>
+        + DataProvider<PatternWhiteSpaceV1Marker>
+        + DataProvider<QuotationMarkV1Marker>
+        + DataProvider<RadicalV1Marker>
+        + DataProvider<RegionalIndicatorV1Marker>
+        + DataProvider<SentenceTerminalV1Marker>
+        + DataProvider<SoftDottedV1Marker>
+        + DataProvider<TerminalPunctuationV1Marker>
+        + DataProvider<UnifiedIdeographV1Marker>
+        + DataProvider<UppercaseV1Marker>
+        + DataProvider<VariationSelectorV1Marker>
+        + DataProvider<WhiteSpaceV1Marker>
+        + DataProvider<XidContinueV1Marker>
+        + DataProvider<GeneralCategoryMaskNameToValueV1Marker>
+        + DataProvider<GeneralCategoryV1Marker>
+        + DataProvider<ScriptNameToValueV1Marker>
+        + DataProvider<ScriptV1Marker>
+        + DataProvider<ScriptWithExtensionsPropertyV1Marker>
+        + DataProvider<XidStartV1Marker>,
+    NP: ?Sized,
+{
+    fn supported_locales(&self) -> Result<Vec<DataLocale>, DataError> {
+        #[allow(clippy::expect_used)]
+        Ok(self
+            .collection
+            .data
+            .read()
+            .expect("poison")
+            .keys()
+            .cloned()
+            .collect::<Vec<_>>())
+    }
+}
+
+struct CompileError {
+    /// offset is the index to an arbitrary byte in the last character in the source that makes sense
+    /// to display as location for the error, e.g., the unexpected character itself or
+    /// for an unknown property name the last character of the name.
+    offset: Option<usize>,
+    /// The type of compile error
+    kind: CompileErrorKind,
+}
+
+impl CompileError {
+    fn explain(self, source: &str) -> DataError {
+        let e = DataError::custom("Invalid transliterator");
+        if let Some(mut col_number) = self.offset {
+            let mut line_number = 1;
+            if let Some(snippet) = source
+                .lines()
+                .filter_map(|line| {
+                    if let Some(snippet) = line.get(col_number..) {
+                        Some(snippet)
+                    } else {
+                        col_number -= line.len();
+                        col_number -= 1; // \n
+                        line_number += 1;
+                        None
+                    }
+                })
+                .next()
+            {
+                e.with_display_context(&format!("at {line_number}:{col_number} '{snippet}'"));
+            }
+        }
+        e.with_debug_context(&self.kind);
+        e
+    }
+}
+
+/// The kind of compile error that occurred.
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+enum CompileErrorKind {
```

Comment:
Thought: Not convinced about moving this error out of the `error` module

Reply:
This type does not exist on the public API, as the `RuleCollection` can only return `DataError`s as a provider. These errors have log information generated from this type's `Debug` implementation, so it really is an implementation detail of `RuleCollection` at the moment.